engine/apps/alerts/models/alert_group.py

```diff
@@ -502,9 +516,58 @@ def declare_incident_link(self) -> str:
     def happened_while_maintenance(self):
         return self.root_alert_group is not None and self.root_alert_group.maintenance_uuid is not None
 
+    @property
+    def response_time(self):
+        """Return time between alert group firing and the first action response, if any."""
+        result = None
+        try:
+            metrics = self.metrics
+        except AlertGroupMetrics.DoesNotExist:
+            pass
+        else:
+            result = metrics.response_time
+        return result
+
+    def _get_response_time(self):
+        """Return response_time based on current alert group status."""
+        response_time = None
+        timestamps = (self.acknowledged_at, self.resolved_at, self.silenced_at, self.wiped_at)
+        min_timestamp = min((ts for ts in timestamps if ts), default=None)
+        if min_timestamp:
+            response_time = min_timestamp - self.started_at
+        return response_time
+
+    def update_response_time(self):
+        """Update stored response time if needed. Only return value if it was set, None otherwise."""
+        updated = None
+        if self.response_time is None:
+            response_time = self._get_response_time()
+            if response_time:
+                metrics, _ = AlertGroupMetrics.objects.update_or_create(
+                    alert_group=self,
+                    defaults={"response_time": response_time},
+                )
+                self.metrics = metrics
+                updated = self.response_time
+        return updated
+
+    def _update_metrics(self, organization_id, previous_state, deleted=False):
+        """Update metrics cache for response time and state as needed."""
+        updated_response_time = self.update_response_time()
+        MetricsCacheManager.metrics_update_cache_for_alert_group(
+            self.channel_id,
+            organization_id=organization_id,
+            old_state=previous_state,
+            new_state=self.state if not deleted else None,
+            response_time=updated_response_time,
+            started_at=self.started_at,
+        )
+
     def acknowledge_by_user(self, user: User, action_source: Optional[str] = None) -> None:
         AlertGroupLogRecord = apps.get_model("alerts", "AlertGroupLogRecord")
+        initial_state = self.state
```

Comment:
Right, I was thinking something more like a context manager, but was leaving for a next iteration (because I was also thinking we may want to do something around the state transitions related to this).